ubuntu_image/classic_builder.py

```diff
@@ -0,0 +1,575 @@
+            # Stop the state machine right here by not appending a next step.
+        else:
+            self._next.append(self.prepare_image)
+
+    def prepare_image(self):
+        # It's required to run ubuntu-image as root to build classic image.
+        check_root_privilege()
+
+        try:
+            # Configure it with environment variables.
+            env = {}
+            if self.args.project is not None:
+                env['PROJECT'] = self.args.project
+            if self.args.suite is not None:
+                env['SUITE'] = self.args.suite
+            if self.args.arch is not None:
+                env['ARCH'] = self.args.arch
+            if self.args.subproject is not None:
+                env['SUBPROJECT'] = self.args.subproject
+            if self.args.subarch is not None:
+                env['SUBARCH'] = self.args.subarch
+            if self.args.with_proposed is not None:
+                env['PROPOSED'] = self.args.with_proposed
+            if self.args.extra_ppas is not None:
+                env['EXTRA_PPAS'] = self.args.extra_ppas
+
+            # Only genereate a single rootfs tree for classic image creation.
+            env['GENERATE_ROOTFS_ONLY'] = '1'
+            live_build(self.unpackdir, env)
+        except CalledProcessError:
+            if self.args.debug:
+                _logger.exception('Full debug traceback follows')
+            self.exitcode = 1
+            # Stop the state machine right here by not appending a next step.
+        else:
+            self._next.append(self.load_gadget_yaml)
+
+    def load_gadget_yaml(self):
+        # The gadget_tree has the same layout as the unpacked gadget snap.
+        # And a piece of gadget.yaml is located at gadget_tree/meta folder.
+        gadget_dir = os.path.join(self.unpackdir, 'gadget')
+        yaml_file = os.path.join(gadget_dir, 'meta', 'gadget.yaml')
+        # Preserve the gadget.yaml in the working dir.
+        shutil.copy(yaml_file, self.workdir)
+        with open(yaml_file, 'r', encoding='utf-8') as fp:
+            self.gadget = parse_yaml(fp)
+        # Make a working subdirectory for every volume we're going to create.
+        # We'll put the volume contents inside these directories, and then use
+        # the directories to create the disk images, one per volume.
+        #
+        # Store some additional metadata on the VolumeSpec object.  This is
+        # convenient, if crufty, since we're poking data onto an object from
+        # the outside.
+        for name, volume in self.gadget.volumes.items():
+            volume.basedir = os.path.join(self.volumedir, name)
+            os.makedirs(volume.basedir)
+        self._next.append(self.populate_rootfs_contents)
+
+    def populate_rootfs_contents(self):
+        src = os.path.join(self.unpackdir, 'chroot')
+        dst = self.rootfs
+        for subdir in os.listdir(src):
+            shutil.move(os.path.join(src, subdir), os.path.join(dst, subdir))
```

Comment:
Since the rootfs has been created with lb running as root, all files there are as root - but they don't have to. The rootfs returned from snap prepare is not root owned and there is nothing wrong with that. So I'd propose we sudo chown the src directory to the running user before we start doing shutil.move(). In my tests this was a very simple:

```
user = os.environ['USER']
run('sudo chown -R {}:{} {}'.format(user, user, src), check=False)
```

But maybe you can find something more sublime. Might be a good idea to put that into a separate method/function that we could easily mock/patch in testing.

The reasons for me proposing these changes are:
1) Currently, if one runs tox locally it will fail on tests that require running the prepare_image() step because we don't have root privileges. It's a bit stupid that we'd have to inform everyone to run tox as sudo as well. I'd prefer running the test suite force us with a password prompt whenever root is required.
2) Would be nice and consistent with what we have now as for xenial builds we get a prompt whenever root is needed anyway.
3) I feel slightly better with the idea or only running with root privileges what needs those privileges.

Reply:
Thanks for your proposal.
`check_root_privilege` is intended to decrease user attention on sudo-prompt for a better user experience. But now, it's a setback for running tests on CI. :face_with_head_bandage: 
I gave several try on u-i with above patch applied and found that u-i asks password twice at times during classic image creation.
But looks like there's no other better solution to workaround the root privilege and make CI feels happy about our tests.
P.S
Another bits need to be changed is how we generate the manifest file. We probably need to chown the manifest file as well there.

Reply:
I played around a cloud image which is generated by ubuntu-image with above patch applied.
The side-effect of above patch is that simply calling chown to change the ownership of rootfs results in the setuid bit lost. Therefore, when we execute some commands with `sudo` after login the system,  we got this
`sudo: /usr/bin/sudo must be owned by uid 0 and have the setuid bit set`
As documented below 
https://www.gnu.org/software/coreutils/manual/html_node/chown-invocation.html#chown-invocation
`The chown command sometimes clears the set-user-ID or set-group-ID permission bits.`
Probably, the above workaround is not ideal in the long run.
I'm trying to find another one to remedy this situation.